### PR TITLE
Override xorm type mapping for U2F counter

### LINKED ISSUE #6168: U2F authentication fails with the message "Could not read your security key. An unknown error occured. Please retry."
- Gitea version (or commit ref): 1.7.1
- Git version: 2.20.1
- Operating system: Debian 9.8 (stretch)
- Database (use `[x]`):
  - [x] PostgreSQL (version 9.6.11)
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist: https://gist.github.com/mporrato/251450f7786075a887ad21a1728a5407

## Description

After upgrading the firmware of my U2F token, Gitea's U2F auth stopped working, showing the message "Could not read your security key. An unknown error occured. Please retry.".

The message in the log seems to refer to the token counter which is correctly defined as `uint32` in the data model in https://github.com/go-gitea/gitea/blob/dfad569e40edf86346a10d11bc68c98c12bca30d/models/u2f.go#L20
Looking at the table in the PostgreSQL database the field is defined as `integer` (which is a 32 bits *signed* value) instead:

```
gitea=# \d u2f_registration 
                                    Table "public.u2f_registration"
    Column    |          Type          |                           Modifiers                           
--------------+------------------------+---------------------------------------------------------------
 id           | bigint                 | not null default nextval('u2f_registration_id_seq'::regclass)
 name         | character varying(255) | 
 user_id      | bigint                 | 
 raw          | bytea                  | 
 counter      | integer                | 
 created_unix | bigint                 | 
 updated_unix | bigint                 | 
Indexes:
    "u2f_registration_pkey" PRIMARY KEY, btree (id)
    "IDX_u2f_registration_created_unix" btree (created_unix)
    "IDX_u2f_registration_updated_unix" btree (updated_unix)
    "IDX_u2f_registration_user_id" btree (user_id)

gitea=# select 4278190365::integer;
ERROR:  integer out of range

```

## Workaround

After manually changing the `counter` column type to `bigint`, Gitea started accepting my U2F token again:

```
gitea=# alter table u2f_registration alter column counter type bigint;
ALTER TABLE
gitea=#
```

This is not a fresh install of 1.7.1 so the problem could be related to the database schema upgrades from previous versions, but from the git blame it looks like the model definition of that table hasn't changed since U2F was introduced in Gitea, so I believe it's more likely to be a problem with the
ORM Gitea is using (sorry, I don't know anything about Gitea internals).

### PR DESCRIPTION
**Pull request by mporrato**
Fix U2F authentication failure with token counter value >= 2^31 when using PostgreSQL.
Closes #6168

### CODE REVIEW

**Comment by jonasfranz** [comment]
Maybe you should consider to add a migration.

**Comment by lafriks** [comment]
Yes, this would need to add migration to update exisiting database

**Review by zeripath — APPROVED** [review]

File: models/migrations/v81.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+	"github.com/go-xorm/xorm"
+)
+
+func changeU2FCounterType(x *xorm.Engine) error {
+	type U2FRegistration struct {
+		Counter uint32 `xorm:"BIGINT"`
+	}
+
+	if err := x.Sync2(new(U2FRegistration)); err != nil {
```
Comment (lafriks) [inline models/migrations/v81.go]: This won't fix existing database. You need to do manual column type change:
For postgres:
```
ALTER TABLE `u2f_registration` ALTER COLUMN `counter` TYPE bigint;
```
For MySQL:
```
ALTER TABLE `u2f_registration` MODIFY COLUMN `counter` BIGINT;
```
Microsoft SQL Server:
```
ALTER TABLE `u2f_registration` ALTER COLUMN `counter` bigint;
```

**EDIT:** No need to do anything for SQLite

~~For SQLite you need to do in transaction:
Rename old table:
ALTER TABLE `u2f_registration` RENAME TO `u2f_registration_old`;
Create new table: `sess.Sync2(new(U2FRegistration))`
Copy data from old to new one:
INSERT INTO `u2f_registration` SELECT * FROM `u2f_registration_old`;
Drop old table
DROP TABLE IF EXISTS `u2f_registration_old`;~~

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
Yesterday, I'm also investigate this bug. My yubikey could work with github but not gitea. I will try this PR to check if it fixed the bug.

**Comment by lunny** [comment]
But it seems this doesn't fix my issue.

**Comment by mporrato** [comment]
@lafriks I believe updating the sqlite3 schema is not needed since it has a single integer type which is big enough to hold an uint32. Sorry for my poor xorm knowledge: haven't used it before. :)

@lunny have you checked if your logs look like https://gist.github.com/mporrato/251450f7786075a887ad21a1728a5407 ? If that's the case, can you connect to the gitea db and do a `\d u2f_registration` ?

File: models/migrations/v81.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	type U2FRegistration struct {
+		Counter uint32 `xorm:"BIGINT"`
+	}
+
+	var err error
+
+	dialect := x.Dialect().DriverName()
+
+	switch dialect {
+	case "mysql":
```
Comment (lafriks) [inline models/migrations/v81.go:19]: You can add:
```
case "tidb":
	fallthrough
```
as MySQL has same syntax.

File: models/migrations/v81.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		_, err = x.Exec("ALTER TABLE u2f_registration MODIFY `counter` BIGINT")
+	case "postgres":
+		_, err = x.Exec("ALTER TABLE u2f_registration ALTER COLUMN \"counter\" SET DATA TYPE bigint")
```
Comment (lafriks) [inline models/migrations/v81.go]: xorm will fix correct esaping
```suggestion
		_, err = x.Exec("ALTER TABLE `u2f_registration` ALTER COLUMN `counter` SET DATA TYPE bigint")
```

File: models/migrations/v81.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	case "tidb":
+		_, err = x.Exec("ALTER TABLE u2f_registration MODIFY `counter` BIGINT")
+	case "mssql":
+		_, err = x.Exec("ALTER TABLE u2f_registration ALTER COLUMN \"counter\" BIGINT")
```
Comment (lafriks) [inline models/migrations/v81.go]: Same here
```suggestion
		_, err = x.Exec("ALTER TABLE `u2f_registration` ALTER COLUMN `counter` BIGINT")
```

File: models/migrations/v81.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		Counter uint32 `xorm:"BIGINT"`
+	}
+
+	var err error
+
+	dialect := x.Dialect().DriverName()
+
+	switch dialect {
+	case "mysql":
+		_, err = x.Exec("ALTER TABLE u2f_registration MODIFY `counter` BIGINT")
```
Comment (lafriks) [inline models/migrations/v81.go]:
```suggestion
		_, err = x.Exec("ALTER TABLE `u2f_registration` MODIFY `counter` BIGINT")
```

File: models/migrations/v81.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	case "postgres":
+		_, err = x.Exec("ALTER TABLE u2f_registration ALTER COLUMN \"counter\" SET DATA TYPE bigint")
+	case "tidb":
+		_, err = x.Exec("ALTER TABLE u2f_registration MODIFY `counter` BIGINT")
+	case "mssql":
+		_, err = x.Exec("ALTER TABLE u2f_registration ALTER COLUMN \"counter\" BIGINT")
+	case "sqlite3":
```
Comment (lafriks) [inline models/migrations/v81.go]: No need for empty case, this can be removed

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lafriks** [comment]
No need to be sorry about, that's why there is review process to learn things :)

**Review by lafriks — APPROVED** [review]

**Comment by techknowlogick** [comment]
I don’t think we need tidb case because for tidb v2 the protocol would be MySQL. Also should sessions be used?

**Comment by lafriks** [comment]
@techknowlogick no session is needed for single sql command

**Comment by hpodhaisky** [comment]
Is this problem still open? I am still getting "Could not read your security key." with  gitea 1.12.1 with sqlite3 on debian stable (buster). My solo key NFC works with github.

**Pull request merged**